ml/algebra/matrix_ops.h

```diff
@@ -155,6 +155,41 @@ diag(const distribution<Float> & d)
 }
 
 template<typename Float>
+boost::multi_array<Float, 2>
+outerProd(const distribution<Float> & d)
+{
+    int dim = d.size();
+    boost::multi_array<Float, 2> D(boost::extents[dim][dim]);
+    for (unsigned i = 0;  i < dim;  ++i)
+    {
+        for (unsigned j = 0;  j < dim;  ++j)
+        {
+            D[i][j] = d[i]*d[j];    
+        }
+    }
+        
+    return D;
+}
+
+template<typename Float>
+boost::multi_array<Float, 2>
+setIdentity(int numDim, boost::multi_array<Float, 2>& D)
+{
+    D.resize(boost::extents[numDim][numDim]);
+    for (unsigned i = 0;  i < numDim;  ++i)
+    {
+        for (unsigned j = 0;  j < numDim;  ++j)
+        {
+            if (i == j)
+                D[i][j] = 1.0f;    
```

Comment:
Float(1) will I believe avoid float-to-double conversions, though it's hardly important.